### PR TITLE
Adding ProvisionedConcurrency Support for Alias

### PR DESCRIPTION
**Pull request by Vandita2020**
Issue #, if available:

**Summary**
Adding support to configure `ProvisionedConcurrency` for alias in ACK.

**Description**
This PR adds support to configure `ProvisionedConcurrency` for Lambda function's alias using ACK. To implement this functionality I added `ProvisionedConcurrencyConfig` as an inline property to alias resource. Thus the user can directly set configurations for Provisioned Concurrency while creating alias. The user can edit the following code to set Provisioned Concurrency along with alias.

```
apiVersion: lambda.services.k8s.aws/v1alpha1
kind: Alias
metadata:
  name: $ALIAS_NAME
  annotations:
    services.k8s.aws/region: $AWS_REGION
spec:
  name: $ALIAS_NAME
  functionName: $FUNCTION_NAME
  functionVersion: $FUNCTION_VERSION
  description: alias created by ACK lambda-controller e2e tests
  provisionedConcurrencyConfig:
    provisionedConcurrentExecutions: $PROVISIONED_CONCURRENT_EXECUTIONS
```
This PR includes both implementation code and e2e tests.

**Acknowledgement**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/resource/alias/hooks.go [hunk 1]
```diff
@@ -70,6 +70,37 @@ func (rm *resourceManager) syncEventInvokeConfig(
...
+		Qualifier:    aws.String(*dspec.Name),
+	}
+
+	if desired.ko.Spec.ProvisionedConcurrencyConfig != nil {
+		if desired.ko.Spec.ProvisionedConcurrencyConfig.ProvisionedConcurrentExecutions != nil {
+			input.ProvisionedConcurrentExecutions = aws.Int64(*desired.ko.Spec.ProvisionedConcurrencyConfig.ProvisionedConcurrentExecutions)
+		} else {
+			input.ProvisionedConcurrentExecutions = aws.Int64(0)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Vandita2020) [reply pkg/resource/alias/hooks.go:94]: I have added the else condition to assign ProvisionedConcurrency value to 0 if user doesn't pass any value.

File: pkg/resource/alias/hooks.go [hunk 2]
```diff
@@ -78,6 +109,21 @@ func (rm *resourceManager) setResourceAdditionalFields(
...
+		ctx,
+		&svcsdk.GetProvisionedConcurrencyConfigInput{
+			FunctionName: ko.Spec.FunctionName,
+			Qualifier:    ko.Spec.Name,
+		},
+	)
+	rm.metrics.RecordAPICall("GET", "GetProvisionedConcurrencyConfig", err)
+	if err != nil {
+		ko.Spec.ProvisionedConcurrencyConfig = nil
+	} else {
```
[the remark this answers is not in the document]
Comment (Vandita2020) [reply pkg/resource/alias/hooks.go:123]: This error is ignored intentionally, the error is returned by the API as there is no existing `provisionedConcurrency` set for the alias, and when it checks the current value in the Console, it doesn't find the configuration, as they are yet not created. So we wanted to ignore that error and instead assign the variable to nil on ACK side before we create it. To make the delta function know that there is a difference in state.

[the remark this answers is not in the document]
Comment (Vandita2020) [reply pkg/resource/alias/hooks.go:123]: Same reason for `FunctionEventInvokeConfig`

File: test/e2e/tests/test_alias.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ def lambda_function():
         replacements["BUCKET_NAME"] = resources.FunctionsBucket.name
         replacements["LAMBDA_ROLE"] = resources.EICRole.arn
         replacements["LAMBDA_FILE_NAME"] = LAMBDA_FUNCTION_FILE_ZIP
-        replacements["RESERVED_CONCURRENT_EXECUTIONS"] = "0"
+        replacements["RESERVED_CONCURRENT_EXECUTIONS"] = "3"
```
[the remark this answers is not in the document]
Comment (Vandita2020) [reply test/e2e/tests/test_alias.py:47]: If we allocate both reserved concurrency and provisioned concurrency for the same function. Then, the amount of provisioned concurrency cannot exceed the amount of reserved concurrency.

File: pkg/resource/alias/hooks.go [hunk 4]
```diff
@@ -78,6 +109,21 @@ func (rm *resourceManager) setResourceAdditionalFields(
 	exit := rlog.Trace("rm.setResourceAdditionalFields")
```
[the remark this answers is not in the document]
Comment (Vandita2020) [reply pkg/resource/alias/hooks.go:109]: Done

File: test/e2e/tests/test_alias.py [hunk 5]
```diff
@@ -196,6 +196,75 @@ def test_smoke_ref(self, lambda_client, lambda_function):
...
+        #Check function_event_invoke_config update fields
+        provisioned_concurrency_config = lambda_validator.get_provisioned_concurrency_config(lambda_function_name,resource_name)
+        assert provisioned_concurrency_config["RequestedProvisionedConcurrentExecutions"] == 2
+
+        # Delete k8s resource
+        _, deleted = k8s.delete_custom_resource(ref)
+        assert deleted
+
+        time.sleep(DELETE_WAIT_AFTER_SECONDS)
+
```
[the remark this answers is not in the document]
Comment (Vandita2020) [reply test/e2e/tests/test_alias.py:264]: The min value of `provisionedConcurrentExecution` needs to be 1, as mentioned here in the [provisionedConcurrencyConfigInput](https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.44.181/service/lambda#PutProvisionedConcurrencyConfigInput) type

File: apis/v1alpha1/alias.go [hunk 6]
```diff
@@ -63,7 +63,8 @@ type AliasSpec struct {
 	FunctionVersion *string `json:"functionVersion"`
 	// The name of the alias.
 	// +kubebuilder:validation:Required
-	Name *string `json:"name"`
+	Name                         *string                               `json:"name"`
+	ProvisionedConcurrencyConfig *PutProvisionedConcurrencyConfigInput `json:"provisionedConcurrencyConfig,omitempty"`
```
[the remark this answers is not in the document]
Comment (Vandita2020) [reply apis/v1alpha1/alias.go:67]: Done

File: pkg/resource/alias/hooks.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (A-Hilaly) [reply pkg/resource/alias/hooks.go:122]: In this case, can we ignore specifically `NotFound` errors? i'm thinking about cases where API returns something different than resource doesn't exist

[the remark this answers is not in the document]
Comment (A-Hilaly) [reply pkg/resource/alias/hooks.go:122]: Can you please share what is the exact error returned by the API?

File: pkg/resource/alias/hooks.go [hunk 7]
```diff
@@ -116,5 +167,27 @@ func (rm *resourceManager) setResourceAdditionalFields(
...
+	if err != nil {
+		return nil
+	}
+
+	err = rm.getFunctionEventInvokeConfig(ctx, ko)
+	if err != nil {
+		return nil
+	}
+
 	return nil
```
Comment (A-Hilaly) [inline pkg/resource/alias/hooks.go:192]: :+1:

File: pkg/resource/alias/hooks.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vandita2020) [reply pkg/resource/alias/hooks.go]: Done

File: templates/hooks/alias/sdk_update_pre_build_request.go.tpl [hunk 8]
```diff
@@ -5,5 +5,15 @@ if delta.DifferentAt("Spec.FunctionEventInvokeConfig"){
...
+
+if delta.DifferentAt("Spec.ProvisionedConcurrencyConfig"){
+    err = rm.updateProvisionedConcurrency(ctx, desired)
+    if err != nil {
+        return nil, err
+    }
+}
+if !delta.DifferentExcept("Spec.ProvisionedConcurrencyConfig"){
     return desired, nil
 }
```
[the remark this answers is not in the document]
Comment (Vandita2020) [reply templates/hooks/alias/sdk_update_pre_build_request.go.tpl]: Writing them together makes sense. We just need an if condition to check the changes if any before executing the rest of the Update function.

**Review by a-hilaly — APPROVED** [review]
Great work on this @Vandita2020 ! :rocket: 
/lgtm

**Pull request merged**